lib/de.ml

```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+      | 0 -> ()
+      | _ ->
+        let len = min block_length 65535 in
+        write_uncompressed_block os len
+          (is_final_block && os.i_pos + len == os.i_len)
+        ; write_uncompressed_blocks os block_length is_final_block
+
+    let flush_output os =
+      if os.o_pos == os.o_len then err_unexpected_end_of_output ()
+      ; while os.bits > 0 do
+          unsafe_set_uint8 os.o os.o_pos os.hold
+          ; os.o_pos <- os.o_pos + 1
+          ; os.bits <- os.bits - 8
+          ; os.hold <- os.hold lsr 8
+        done
+      ; os.o_pos
+
+    let compress_none _c i o =
+      let os = init_output i o in
+      write_uncompressed_blocks os os.o_len true
+      ; flush_output os
+
+    let precode_lens_permutation =
+      [|16; 17; 18; 0; 8; 7; 9; 6; 10; 5; 11; 4; 12; 3; 13; 2; 14; 1; 15|]
+
+    let compute_precode_items lens num_lens precode_freqs precode_items =
+      Array.fill precode_freqs 0 (Array.length precode_freqs) 0
+
+      ; let itemptr = ref 0 in
+        let run_start = ref 0 in
+        let rec f () =
+          let len = lens.(!run_start) in
+          let run_end = ref !run_start in
+          while !run_end <> num_lens && len == lens.(!run_end) do
+            incr run_end
+          done
+          ; if len == 0 then begin
+              while !run_end - !run_start >= 11 do
+                let extra_bits = min (!run_end - !run_start - 11) 0x7F in
+                precode_freqs.(18) <- precode_freqs.(18) + 1
+                ; precode_items.(!itemptr) <- 18 lor (extra_bits lsl 5)
+                ; incr itemptr
+                ; run_start := !run_start + 11 + extra_bits
+              done
+              ; if !run_end - !run_start >= 3 then (
+                  let extra_bits = min (!run_end - !run_start - 3) 0x7 in
+                  precode_freqs.(17) <- precode_freqs.(17) + 1
+                  ; precode_items.(!itemptr) <- 17 lor (extra_bits lsl 5)
+                  ; incr itemptr
+                  ; run_start := !run_start + 3 + extra_bits)
+            end
+            else if !run_end - !run_start >= 4 then (
+              precode_freqs.(len) <- precode_freqs.(len) + 1
+              ; precode_items.(!itemptr) <- len
+              ; incr itemptr
+              ; incr run_start
+              ; while !run_end - !run_start >= 3 do
+                  let extra_bits = min (!run_end - !run_start - 3) 0x3 in
+                  precode_freqs.(16) <- precode_freqs.(16) + 1
+                  ; precode_items.(!itemptr) <- 16 lor (extra_bits lsl 5)
+                  ; incr itemptr
+                  ; run_start := !run_start + 3 + extra_bits
+                done)
+          ; while !run_start <> !run_end do
+              precode_freqs.(len) <- precode_freqs.(len) + 1
+              ; precode_items.(!itemptr) <- len
+              ; incr itemptr
+              ; incr run_start
+            done
+          ; if !run_start <> num_lens then f () else !itemptr in
+        f ()
```

Comment:
Not sure that the best should be to use a recursive function when you use reference inside, again two possibilities exist:
- re-implement it in a full functional way (without any reference)
- use a loop with reference

The mix of both (recursive function & reference) is hard to optimize then.

Reply:
Indeed, those are relics of when I was unsure about which one to choose, for now, I'll stick to loops